### ISSUE TITLE
LM4F: Doesn't compile in C++ due to non-constexpr enum values in systemcontrol.h

### ISSUE THREAD

**Issue by ktims** [libopencm3/libopencm3#1385]
`lm4f_clken` in `systemcontrol.h` describes a number of enum values with the form:

```RCC_WD0 = ((uint32_t)&SYSCTL_RCGCWD - SYSCTL_BASE) << 5```

In C++, casting `(volatile uint32_t*)` (which is the type of SYSCTL_RCGCWD due to MMIO32) to `uint32_t` ends up in a call to `reinterpret_cast` which is not `constexpr`. Therefore, these values are not valid in the definition of an `enum`. As GCC explains:

```
/home/ktims/.platformio/packages/framework-libopencm3/include/libopencm3/lm4f/systemcontrol.h:493:23: error: 'reinterpret_cast<volatile uint32_t* {aka volatile long unsigned int*}>(1074783744)' is not a constant expression
  493 |  RCC_WD0 = ((uint32_t)&SYSCTL_RCGCWD - SYSCTL_BASE) << 5,
/home/ktims/.platformio/packages/framework-libopencm3/include/libopencm3/lm4f/systemcontrol.h:493:56: error: enumerator value for 'RCC_WD0' is not an integer constant
  493 |  RCC_WD0 = ((uint32_t)&SYSCTL_RCGCWD - SYSCTL_BASE) << 5,
```

I don't think there is a clean way to cast pointer to integer that *is* `constexpr`, so perhaps the cleanest approach is to provide integer-type defines as well as pointer-type defines that are based on them.

**Comment by karlp** [comment]
hey neat, someone using lm4f :)

IMO, the pointer _is_ const, and this is c++'s fault, but I realise that doesn't help anyone much :(  I mean, you can see right there that it's a literal number.  I'm certainly against having two definitions, but if you come up with a new definition that works, I'll happily look at it.

I _suspect_ this would work with just using the _REG_BIT() style from stm32, (I think, iirc, that stm32 actually got the idea _from_ lmf4 even, but it's been improved a bit I guess)

Have a look at https://github.com/libopencm3/libopencm3/blob/master/include/libopencm3/stm32/g4/rcc.h#L74-L75 and also https://github.com/libopencm3/libopencm3/blob/master/include/libopencm3/stm32/g4/rcc.h#L778 and onwards.

If you can convert the lm4f systemcontrol to something like that, does that fix your constexpr issues? because I'm 99.9% sure people are using c++ with the stm32 parts at least.

**Comment by ktims** [comment]
> hey neat, someone using lm4f :)

The times we're living in, eh? :p

> IMO, the pointer is const, and this is c++'s fault, but I realise that doesn't help anyone much :( I mean, you can see right there that it's a literal number. I'm certainly against having two definitions, but if you come up with a new definition that works, I'll happily look at it.

I wouldn't argue, but the compiler has to follow the language specification, and C-style casts turn into `reinterpret_cast`, which isn't `constexpr` due to UB. I'd think it should be able to prove its const-ness because after macro expansion it's a literal, but I guess that 'literalness' doesn't propagate. Or maybe the compiler doesn't care about const-ness and it's explicitly forbidden by the standard.

> If you can convert the lm4f systemcontrol to something like that, does that fix your constexpr issues? because I'm 99.9% sure people are using c++ with the stm32 parts at least.

STM32 works fine in C++ yeah, and that pattern is what I was suggesting.  Something like:

```
#define SYSCTL_RCGCWD_OFFSET (0x600)
#define SYSCTL_RCGCWD MMIO32(SYSCTL_BASE + SYSCTL_RCGCWD_OFFSET)
...
RCC_WD0 = (SYSCTL_RCGCWD_OFFSET << 5),
```

For now I've just patched the literal values (eg 0x600) directly into the enum definitions, but if a PR would be accepted I can work on one.

**Comment by karlp** [comment]
that sounds fine, by all means send that in, c++ being weird or not, we're definitely not trying to _stop_ people from using it :)
